sharding-proxy/sharding-proxy-frontend/sharding-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/shardingproxy/frontend/mysql/auth/MySQLAuthenticationEngine.java

```diff
@@ -54,20 +56,31 @@ public void handshake(final ChannelHandlerContext context, final BackendConnecti
     @Override
     public boolean auth(final ChannelHandlerContext context, final PacketPayload payload, final BackendConnection backendConnection) {
         MySQLHandshakeResponse41Packet response41 = new MySQLHandshakeResponse41Packet((MySQLPacketPayload) payload);
-        if (authenticationHandler.login(response41.getUsername(), response41.getAuthResponse())) {
-            if (!Strings.isNullOrEmpty(response41.getDatabase()) && !LogicSchemas.getInstance().schemaExists(response41.getDatabase())) {
+        final Optional<ProxyUser> user = authenticationHandler.login(response41.getUsername(), response41.getAuthResponse());
+        if (!user.isPresent()) {
```

Comment:
**Hi, Haisheng, i appreciate your bugfix and  unit tests!**
Here is my suggestion, can you consider it is better or not?

> How about move the judegment of user existence and its authorized schemas to `authenticationHandler.login()`, therefore `AuthenticationHandler` will just tell `MySQLAuthenticationEngine` that the login information is true or not. And the task of `MySQLAuthenticationEngine` is only to set properties for `backendConnection` or call `context.writeAndFlush(getAccessDeniedError(context, response41))` by login information.

Reply:
OK 👍